### ISSUE TITLE
fixing ^(a::perm, n::Int)

### ISSUE THREAD

**Issue by kalmarek** [Nemocas/Nemo.jl#167]
currently, a little unexpectedly:
* for `n >= 2`  it works (albeit not efficiently, we could compute order by cycle decomposition)
* for `n = 1` it fails with `MethodError: no method matching copy(::Nemo.perm)`
* for `n = 0` it fails with `MethodError: no method matching one(::Nemo.perm)`
* for `n < 0` it fails with familiar (though unexpected) `ERROR: DomainError: Cannot raise an integer x to a negative power -n`.

Solutions:
1. define `power_by_squaring(p::perm, n::Int)` to which `^` defers.
2. define `copy` (similar to `deepcopy_internal`) and `one`; (user handles negative powers by `inv(p)^n`

This also brings the issue of `one` vs. `eye` for Groups;
Has anyone an idea how to handle e.g. Rings -> Groups forgetful functors, i.e. multiplicative `one` vs. additive `zero`?

3. Simple 
```julia
function (^)(a::perm, n::Int)
   if n < 0
      return inv(a)^-n
   elseif n == 0
      return parent(a)()
   elseif n == 1
      return a
   else
      return Base.power_by_squaring(a, n % order(a))
   end
end

copy(p::perm) = deepcopy(p)
```

**Comment by thofma** [comment]
At the moment we don't make use of `Base.power_by_squaring` and I don't think we should. It think it's best to just write a custom `^(::perm, ::Int)`, see https://github.com/Nemocas/Nemo.jl/blob/master/src/generic/Matrix.jl#L540-L564
for a fast non-recursive implementation. The use of `order` sounds like a good idea. But if you already have the cycle decomposition to compute the order, why not use it to compute the power?

Regarding `copy`: I don't think copy should default to `deepcopy`.

Regarding `one` and `eye`. I don't think `eye` is a good name in the context of (permutation) groups. Actually it is not a good name in the context of matrices either. It's a confusing MATLAB anachronism. If you don't come from MATLAB, it is impossible to guess the right name for constructing an identity matrix. Now for the (permutation) groups we should do something like `id(::PermGrp)` or `identity(::PermGrp)`. This is much more intuitive.

**Comment by kalmarek** [comment]
Will do `^(a::perm, n::Int)` as You suggest (using cycles);
the implementation You linked throws on `n<0`. Is this a common practice across Nemo?
(`MatElem` can be non-invertible, but it could be `inv` throwing).
For groups it makes sense to allow `n<0` and return n-th power of `inv`. 

`copy` : where `Base.power_by_squaring` uses `copy`, the linked source uses `deepcopy`;
this definition was only to make `Base.power_by_squaring` work, anyway.

As I don't come from matlab I fully agree on the point; `identity` is already a Julia (identity) function;
`id` seems to be a better choice then;
Should we define such methods for specific groups (like `PermGroup`) or more generally `id{T<:Group}(G::T) = G()`??

This is also a question on extracting a group operation and forming a group from Rings, etc.
I'm tackling `DirectProductGroup` (or should we consider `{T<:Group}DirectProduct{T}`??) locally and it seems natural to include some mechanism of obtaining the group operation and the group inverse from a Ring (right now I assume it's `+`);
This is probably topic on a different discussion;

**Comment by wbhart** [comment]
On 13 May 2017 at 23:40, kalmarek <notifications@github.com> wrote:

> Will do ^(a::perm, n::Int) as You suggest (using cycles);
> the implementation You linked throws on n<0. Is this a common practice
> across Nemo?
> (MatElem can be non-invertible, but it could be inv throwing).
> For groups it makes sense to allow n<0 and return n-th power of inv.
>

Nemo does this quite a lot in other places, and it should be the case
everywhere that inversion makes sense. Sometimes it throws instead because
of issues with the underlying implementation that haven't been fixed, I
think.


> This is also a question on extracting a group operation and forming a
> group from Rings, etc.
>
Most people would think of the additive group of a ring as the group
associated with the ring, but would expect the identity of a ring to be 1.
So I don't think this should be how rings are handled.

As for groups, I think it is going to matter whether they are additive or
multiplicative (where it makes sense to make such a distinction). Whilst we
could just define G() to always return the identity, that will be confusing
for rings, where it always returns 0, as you noted. I have no particular
opinion on this, so I don't mind whatever is decided.

If it is superconvenient just to define G() to always return the identity
of a group, just do it.

> I'm tackling DirectProductGroup (or should we consider
> {T<:Group}DirectProduct{T}??)
>
I think parameterising DirectProduct might makes some sense. But there are
a few things to consider.

Most groups are going to be complicated enough that you need a type for the
parent object and a type for the elements in the group itself. Then, the
direct product will probably be represented by an array of _elements_ in
the group(s), and certainly not an array of groups.

In the case of a direct power for example, you might get away with
representing it as an array of elements of all the same type. The type for
the direct power will then need to be parameterised by the type of the
elements of the array, i.e. by the type of group elements, not the type of
group.

This is what happened with polynomials, where I needed an array of
coefficients of a given type to represent a polynomial. Initially I was
trying to parameterise polynomials by the type of the coefficient ring
parent object, but later realised I had to parameterise by the type of the
_elements_ in the coefficient ring.

This is why there are both Group and GroupElem abstract types in Nemo. One
is for the types of the group parent objects, the other is for the types of
the group elements.

Eventually, you encounter objects that can be thought of as both parent
objects and elements. For example, ideals can be thought of as elements of
a set of ideals, or as a parent for the elements of the ideal.

It can get even more complicated. In the case of ideals over a polynomial
ring over a coefficient ring, it made sense to parameterise the ideals by
the type of the elements in the coefficient ring, not by the type of the
elements of the polynomial ring. So there's no hard and fast rule. It's
whatever i actually needed by the implementation.

For the time being though, I recommend parameterising on the type of the
elements of the underlying group, since this is probably the parameter you
actually need to define the type.

The more important consideration is whether you need a parameter at all. If
you can omit a parameter, it is worth doing. But very often they are
actually needed to make the type work.

Direct products are of course complicated because they might be a direct
product of multiple different groups. One could in theory parameterise them
by a tuple of element types, one for each group (though this is not easy to
do). But this means there will be a different type for every different kind
of direct product. That might be overkill. The JIT compiler has to
recompile functions every time the type changes, so this could lead to long
delays if many different direct products occur in a computation.

It might be better to store an array of groups in the parent object for a
direct product, rather than try to parameterise the direct product types.

You can do things like

type DirectProductGroup <: Group
   base_groups::Array{Group, 1}
end

type DirectProduct <: GroupElem
   A::Array{GroupElem, 1}
   parent::DirectProductGroup
end

This will allow you to represent a direct product (element) by an array of
group elements. Behind the scenes, this is pretty much equivalent to an
array of Any, which really gets interpreted rather than compiled. But it
has the additional safety of only allowing elements to be inserted in the
array if they can be converted to group elements. So it wouldn't accept a
double precision float, for example. It's no faster than A::Array{Any, 1},
but it's safer.

Note that this approach only requires a new parent object for each kind of
direct product. Functions that apply to direct products don't need to be
recompiled, as the type doesn't vary for each different kind of direct
product.

You might also consider whether one really needs the pointer to the parent
field in the DirectProduct type, since it can be computed if it is ever
needed, by calling a function called parent. Technically there is enough
information in that array of GroupElem's to compute the parent as needed.

The decisive factor is usually whether you want constructors such as

direct_product(g1, g2, g3, g4)

where g1, g2, g3 and g4 are elements of existing groups, or whether you
want to require the user to have to construct the parent object before
creating direct product elemrents, e.g:

G = DirectProduct(G1, G2, G3, G4) # construct the parent object first
g = G([g1, g2, g3, g4]) # now use it to construct elements of that direct
product

For example, elsewhere in Nemo, we originally set it up so that the user
had to create a matrix space before creating a matrix in that matrix space.
(A matrix space is the parent object of a matrix.) But we've now decided we
want to be able to create matrices directly and have the matrix space
computed as needed when the user calls the parent function on a given
matrix. A similar thing happens when creating ideals, and possibly also
quotient rings.

In short, there's no one way to do it. But some patterns are emerging in
what is actually convenient (and efficient).

**Comment by kalmarek** [comment]
@wbhart thanks for the thorough explanation;
have a look at my current implementation here: https://gist.github.com/kalmarek/00968aaf4fac0deac65043da08a3707c

I was wandering about something like DirectProduct{some Nemo abstract type}, to that the constructor would know which category I want my product belong to (e.g. Groups, Rings...) and would do the right conversion of the individual factors, etc;

as You can see for now i settled on simple DirectProduct{Group/GroupElem} cause this was the simplest to get my work going;

**Comment by wbhart** [comment]
I'm a bit confused how the current implementation will work. You have a
direct product element type, but these seem to consist of a vector of
groups. Are you trying to construct things like (g_1, g_2, ..., g_n) where
g_i \in G_i for groups G_i (probably all the same group in your current
implementation), or something like G_1 x G_2 x ... x G_n. If you want the
latter, how are you representing elements of that group? If you want the
former, I think you want a vector of GroupElem's.

**Comment by wbhart** [comment]
Also, you might not realise that if you do T <: Group (where you probably
mean T <: GroupElem), this restricts T to always being the same type. In
other words, you won't be able to construct direct products of different
kinds of groups.

Instead of parameterising the type, you just want Vector{GroupElem}
directly, I think. This allows you to construct a vector out of any group
elements from any groups.

**Comment by kalmarek** [comment]
Yeah, that was a non-working version; the working has
```julia
type DirectProductGroup <: Group
   factors::Vector{Group}
   operations::Vector{Function}
end

type DirectProductGroupElem <: GroupElem
   elts::Vector{GroupElem}
   parent::DirectProductGroup

   DirectProductGroupElem{T<:GroupElem}(a::Vector{T}) = new(a)
end
```
as You suggested.

**Comment by thofma** [comment]
Note that `Vector{GroupElem}` will be very slow, since the entries the element type of this array is abstract.

The way to go should be tuples, not arrays. Maybe `DirectGroup(::Group, ::Int)` using homogenous `NTuple`s and then non-homogenous using arbitrary `Tuple`s.

Don't use arrays for this.

**Comment by kalmarek** [comment]
I will have to look into using tuples (which seem more natural in the context of direct products, actually).

**Comment by wbhart** [comment]
Tommy, I agree that will be faster. Very hard to implement though.

Also, can you imagine an algorithm that needs to print out every possible
tuple of elements from say ten distinct groups? It would need to compile
1024 times.

I ultimately don't mind how it's done, and my example is a bit extreme. But
it's not totally clear you can avoid arrays here.

**Comment by thofma** [comment]
Not sure I understand. You only get 1024 many compilations if the ten groups have different types. If they are all of type `PermGroup`, there is only one compilation for `DirectProductGroup{PermGroup, PermGroup}`.

@kalmarek Do you already need rings in your application of direct products? What kind of rings do you want to use?

I am not sure how "rings as groups" part should work in Nemo.

**Comment by wbhart** [comment]
I'm referring to the types of the tuples, which would all be different.

**Comment by kalmarek** [comment]
@thofma 
e.g. I use `FiniteField(p,1,"x")` as substitute for finite cyclic groups;
The question about "Rings as groups" is more of a question if there is a canonical way of doing this which I have missed. 
Right now I'm doing group rings for the Wreath Products of `Z/p` and `S_n` and "additive groups of rings" are surely hackable in this context. 
For this application elements (of the embedded direct product) parametrised by `elem_type(Z/p)` would suffice; 
However the groups are pretty small (a few thousands of elements) so I was focused not so much on the performance but rather to get things going; It'd be nice to get it right, though;

@wbhart thanks again for the 'parametrise by type of coefficients' tip; it helped me greatly with my implementation of group rings; If You have time and patience You can have a look at my efforts here: https://git.wmi.amu.edu.pl/kalmar/GroupRings.jl/src/enh/Nemoification

**Comment by thofma** [comment]
OK. Getting the "rings as groups" thing right is not trivial.

I have also looked at your group rings. With your current implementation you cannot use Nemo integers. You need `{T <: RingElem}` in the type definition of `GroupRingElem`. Or just use `{T}` if you also want to allow julia number types like `BigInt`.

Also your type `GroupRing` needs two parameters if you care about performance:
```julia
type GroupRing{G, E} <: Ring
   group::G
   basis::Vector{E}
   basis_dict::Dict{E, Int}
   pm::Array{Int,2}
end
```
If I understand it correctly, you only do the integral group ring, right? Maybe you should call it `IntegralGroupRing` or something like that.

**Comment by wbhart** [comment]
The Nemoification is looking good. Here are some additional comments.

It's better to put the types in a separate file. Julia cannot forward
reference types, and so eventually you need all the types defined before
any functions. If you put all the types in a separate file, you can include
them before any functions.

Once you parameterise GroupRIng in the way Tommy suggests, you should
probably also not use plain Vector as type in some of your functions.
Instead you should probably restrict to Vector{T} for some kind of T, e.g.
T <: GroupElem, unless you really want this to be anything at all.

I could be wrong, but I don't think you need parentheses around the
function names for operators like (==), (+), (-), etc.

I don't think it is a good idea to define (*)(X::GroupRingElem,
Y::GroupRingElem)
= groupring_mult(X,Y). Instead, just call the other functions *. The
groupring_mul! function can just be called mul!. It's never a good idea to
put the name of a type in a function name in Julia. It usually means you
aren't using the multiple dispatch properly.

You might need to export star and a few other functions that you want the
user to have access to.

Unfortunately Tommy is right. If you want people to be able to use Nemo
integers, you can't do T <: Number at present. Just do {T}. Later when
traits finally get added to Julia, we might be able to do something about
this.

**Comment by kalmarek** [comment]
Thanks for the suggestions; will try to apply them; This is a rewrite that I have not put into production ;-) I will be more careful about types in the future.

@thofma I mostly work in `QQ[G]`, `RR[G]` and `RealIntervalField[G]`. There is nothing integral in the definition of `GroupRing` -- It may feel strange, but that was what @wbhart suggested and made the nemoification work: `GroupRingElems` are typed by the type of coefficients; `GroupRing` objects are completely oblivious to coefficients.

Yes, You don't need parenthesis around operators.
Point about multiple dispatch -- taken;

So far I didn't need to use Nemo integers/etc for coefficients. Aside from beeing more general, what might be the advantage? E.g. are flint rationals faster than native julias ones?

> For example, elsewhere in Nemo, we originally set it up so that the user
> had to create a matrix space before creating a matrix in that matrix space.
> (A matrix space is the parent object of a matrix.) But we've now decided we
> want to be able to create matrices directly and have the matrix space
> computed as needed when the user calls the parent function on a given
> matrix. A similar thing happens when creating ideals, and possibly also
> quotient rings.

Do You have an example in code of such lazy initialisation of parent?

**Comment by wbhart** [comment]
On 22 May 2017 at 16:21, kalmarek <notifications@github.com> wrote:

> Thanks for the suggestions; will try to apply them; This is a rewrite that
> I have not put into production ;-) I will be more careful about types in
> the future.
>
> @thofma <https://github.com/thofma> I mostly work in QQ[G], RR[G] and
> RealIntervalField[G]. There is nothing integral in the definition of
> GroupRing -- It may feel strange, but that was what @wbhart
> <https://github.com/wbhart> suggested and made the nemoification work:
> GroupRingElems are typed by the type of coefficients; GroupRing objects
> are completely oblivious to coefficients.
>
> Yes, You don't need parenthesis around operators.
> Point about multiple dispatch -- taken;
>
> So far I didn't need to use Nemo integers/etc for coefficients. Aside from
> beeing more general, what might be the advantage? E.g. are flint rationals
> faster than native julias ones?
>

No, you are essentially doing the right thing in keeping it independent of
Flint. However, users of Nemo.jl might actually want to work with Nemo
integers without having to convert them to Julia ones first. One way of
doing that (the only way at the moment) is to accept a general type
parameter T rather than use T <: Number. This won't slow your code down at
all, but may make some Nemo users happy.

It's up to you of course, but eventually someone is going to ask for this.

> For example, elsewhere in Nemo, we originally set it up so that the user
> had to create a matrix space before creating a matrix in that matrix space.
> (A matrix space is the parent object of a matrix.) But we've now decided we
> want to be able to create matrices directly and have the matrix space
> computed as needed when the user calls the parent function on a given
> matrix. A similar thing happens when creating ideals, and possibly also
> quotient rings.
>
> Do You have an example in code of such lazy initialisation of parent?
>

I think ideals in Singular.jl is the only place I can point to. See line
201 and following in
https://github.com/wbhart/Singular.jl/blob/master/src/ideal/ideal.jl

I don't think we've actually switched matrices over fully yet. But we will,
eventually. It's a nuisance having to create the parent object before you
can do anything.

We've gotten part of the way there for matrices, by overloading Julia's
hvcat functions, etc, see line 3454 and following in
https://github.com/Nemocas/Nemo.jl/blob/master/src/generic/Matrix.jl

This allows one to create matrices using R[1 2 3; 4 5 6] where R is a Nemo
parent object for the *coefficient ring*, not the matrix space over that
coefficient ring.

Matrices are a special case, of course, since they have this nice syntax in
Julia, which also happens to work just fine with Nemo (coefficient ring)
parent objects, just as easily as it works for Julia types.

Bill.

**Comment by thofma** [comment]
> @thofma I mostly work in QQ[G], RR[G] and RealIntervalField[G]. There is nothing integral in the definition of GroupRing -- It may feel strange, but that was what @wbhart suggested and made the nemoification work: GroupRingElems are typed by the type of coefficients; GroupRing objects are completely oblivious to coefficients.

I don't think that is what he suggested.

**Comment by kalmarek** [comment]
fixed by https://github.com/Nemocas/Nemo.jl/pull/170
